File: Iteration 1/Source Code/Course.java
import java.util.ArrayList;

public class Course {
    private String courseId;
    private String courseName;
    private Lecturer lecturer;
    private ArrayList<Student> students;
    private ArrayList<CourseSection> courseSections;
    private Course prerequisiteCourse;
    private String prerequisiteID;
    private int credits;

    public Course() {
    }

    public Course(String courseId, String courseName) {
        this.courseId = courseId;
        this.courseName = courseName;
    }
    
    public Course(String courseId, String courseName, int credits, String prerequisiteID) {
        this.courseId = courseId;
        this.courseName = courseName;
        this.credits = credits;
        this.prerequisiteID=prerequisiteID;
    }


    public String getCourseId() {
        return courseId;
    }

    public String getCourseName() {
        return courseName;
    }

    public Lecturer getLecturer() {
        return lecturer;
    }

    public void addStudent(Student student) {
        try {
            if (student != null)
                students.add(student);
            else
                System.out.println("The student object is null");
        } catch (NullPointerException e) {
            System.out.println("The students list has not been initialized.");
        } catch (Exception e1) {
            System.out.println(e1.getMessage());
        }
    }

    public Course getPrerequisiteCourse() {
        return prerequisiteCourse;
    }

    public void setPrerequisiteCourse(Course prerequisiteCourse) {
        this.prerequisiteCourse = prerequisiteCourse;
    }

    public String getPrerequisiteID() {
        return prerequisiteID;
    }

    public int getCredits() {
        return credits;
    }
}

File: Iteration_3/Course.py
from typing import List, Optional, TYPE_CHECKING
from Logging_Config import logger
if TYPE_CHECKING:
    from Student import Student
    from CourseSection import CourseSection
    from Lecturer import Lecturer

class Course:
    def __init__(self,
                 course_id: str,
                 course_name: str,
                 credits: int = 0,
                 prerequisite_course: Optional['Course'] = None,
                 semester: int = 0,
                 grade: Optional[str] = None):
        self.__course_id = course_id
        self.__course_name = course_name
        self.__credits = credits
        self.__prerequisite_course = prerequisite_course
        self.__semester = semester
        self.__grade = grade
        self.__lecturer: Optional['Lecturer'] = None
        self.__students: List['Student'] = []
        self.__course_sections: List['CourseSection'] = []

    def get_course_type(self) -> str:
        raise NotImplementedError("Subclasses must implement get_course_type()")

    def get_course_id(self) -> str:
        return self.__course_id

    def get_course_name(self) -> str:
        return self.__course_name

    def get_lecturer(self) -> Optional['Lecturer']:
        return self.__lecturer

    def get_grade(self) -> Optional[str]:
        return self.__grade
    
    def set_grade(self, grade: str) -> None:
        self.__grade = grade

    def add_student(self, student: 'Student') -> None:
        try:
            if student is not None:
                self.__students.append(student)
            else:
                logger.error("The student object is null")
        except AttributeError:
            logger.error("The students list has not been initialized.")
        except Exception as e:
            logger.error(str(e))

    def get_prerequisite_course(self) -> Optional['Course']:
        return self.__prerequisite_course

    def set_prerequisite_course(self, prerequisite_course: 'Course') -> None:
        self.__prerequisite_course = prerequisite_course

    def get_credits(self) -> int:
        return self.__credits

    def get_students(self) -> List['Student']:
        return self.__students

    def set_students(self, students: List['Student']) -> None:
        self.__students = students

    def get_course_sections(self) -> List['CourseSection']:
        return self.__course_sections

    def set_course_sections(self, course_sections: List['CourseSection']) -> None:
        self.__course_sections = course_sections

    def get_semester(self) -> int:
        return self.__semester

    def __str__(self) -> str:
        return f"{self.__course_name} {self.__course_id}"
    
    def get_credits(self) -> int:
        return self.__credits